homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,360 @@
+    "SERVICE_INSTALL",
+    "SERVICE_SKIP",
+    "SUPPORT_BACKUP",
+    "SUPPORT_INSTALL",
+    "SUPPORT_PROGRESS",
+    "SUPPORT_SPECIFIC_VERSION",
+    "UpdateDeviceClass",
+    "UpdateEntity",
+    "UpdateEntityDescription",
+]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Select entities."""
+    component = hass.data[DOMAIN] = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_INSTALL,
+        {
+            vol.Optional(ATTR_VERSION): cv.string,
+            vol.Optional(ATTR_BACKUP): cv.boolean,
+        },
+        async_install,
+    )
+
+    component.async_register_entity_service(
+        SERVICE_SKIP,
+        {},
+        UpdateEntity.async_skip.__name__,
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+async def async_install(entity: UpdateEntity, service_call: ServiceCall) -> None:
+    """Service call wrapper to validate the call."""
+    # This entity does not support installing updates.
+    if not entity.supported_features & SUPPORT_INSTALL:
+        raise HomeAssistantError(
+            f"Installing updates is not supported for {entity.name}"
+        )
+
+    # If version is not specified, but no update is available.
+    if (version := service_call.data.get(ATTR_VERSION)) is None and (
+        entity.current_version == entity.latest_version or entity.latest_version is None
+    ):
+        raise HomeAssistantError(f"No update available for {entity.name}")
+
+    # If version is specified, but not supported by the entity.
+    if version is not None and not entity.supported_features & SUPPORT_SPECIFIC_VERSION:
+        raise HomeAssistantError(
+            f"Installing a specific version is not supported for {entity.name}"
+        )
+
+    # If backup is requested, but not supported by the entity.
+    if (
+        backup := service_call.data.get(ATTR_BACKUP)
+    ) and not entity.supported_features & SUPPORT_BACKUP:
+        raise HomeAssistantError(f"Backup is not supported for {entity.name}")
+
+    # Update is already in progress.
+    if entity.in_progress is not False:
+        raise HomeAssistantError(
+            f"Update installation already in progress for {entity.name}"
+        )
+
+    await entity.async_install_with_progress(version, backup)
+
+
+@dataclass
+class UpdateEntityDescription(EntityDescription):
+    """A class that describes update entities."""
+
+    device_class: UpdateDeviceClass | str | None = None
```

Comment:
Should we allow `str` ?

Reply:
Well is supports device classes, so custom device classes for state translations can be provided, which ar etrings.